hypothesis-python/docs/details.rst

```diff
@@ -694,3 +694,44 @@ on working with markers <pytest:mark examples>`.
 .. note::
     Pytest will load the plugin automatically if Hypothesis is installed.
     You don't need to do anything at all to use it.
+
+
+.. _fuzz_one_input:
+
+-------------------------
+Use with external fuzzers
+-------------------------
+
+Sometimes, you might want to point a traditional fuzzer such as
+`python-afl <https://github.com/jwilk/python-afl>`__ or :pypi:`pythonfuzz`
+at your code. Wouldn't it be nice if you could use any of your
+:func:`@given <hypothesis.given>` tests as fuzz targets?
+
+.. code:: python
+
+    @given(st.text())
+    def test_foo(s):
+        ...
+
+
+    # This is a traditional fuzz target - call it with a bytestring,
+    # or a binary IO object, and it runs the test once.
+    fuzz_target = test_foo.hypothesis.fuzz_one_input
+
+    # For example:
+    fuzz_target(b"\x00\x00\x00\x00\x00\x00\x00\x00")
+    fuzz_target(io.BytesIO(...))
+
+If the bytestring was invalid, for example because it was too short or
+failed a filter or :func:`~hypothesis.assume` too many times,
+``fuzz_one_input`` returns ``None``.
+
+If the bytestring was valid and the test passed, ``fuzz_one_input`` returns
+a canonicalised and pruned buffer which will replay that test case.  This
+is provided as an option to improve the performance of mutating fuzzers,
+but can safely be ignored.
+
+If the test *failed*, i.e. raised an exception, ``fuzz_one_input`` will add
```

Comment:
I'm not totally wild about this interface. In particular it means that a fuzzer running a Hypothesis-based test will never report any crashes.

Realistically I think that nobody is going to use the canonicalised information and that we'd be better off behaving more like a traditional fuzzer target, so my recommendation would be to keep saving it in the example database but also to raise the error, and to not return the canonicalized input.

Reply:
Wait, sorry, I misread what the code was doing. It *does* do that. Disregard the above comment, but also note that the documentation confused me and should maybe be clearer!

Also I'm still not sure returning the canonicalized input is that useful.

Reply:
I'll clarify the docs.

Returning canonicalised input may not be useful very often, but it's so easy that I thought we might as well.  If nothing else it makes it easy to generate a random corpus for mutating fuzzers...

Reply:
My general bias is that "it's easy to do and might be useful" is not a good reason to do it in the absence of a more compelling reason to do something, because the real cost is not the performance but that it makes it something we have to support later.

If there turns out to be a compelling use case for it I'm more than happy for that, but I just don't think any of the available fuzzers are going to be able to take advantage of it, and I think any fuzzer we write is going to want a more sophisticated interface that uses the full `ConjectureResult` structure.

Reply:
FWIW I've been using it while playing with pythonfuzz - basically a pure-python afl knockoff, but getting the canonical buffer (especially autopruning) made crossover mutation operators considerably more useful.

So it's not exactly out-of-the-box compatible, but IMO sufficient to justify the feature.